src/canisters/frontend/ic-certified-assets/src/lib.rs

```diff
@@ -452,20 +369,301 @@ pub fn post_upgrade(stable_state: StableState, args: Option<AssetCanisterArgs>)
         set_permissions
     });
 
-    STATE.with(|s| {
-        *s.borrow_mut() = State::from(stable_state);
-        set_certified_data(&s.borrow().root_hash());
+    with_state_mut(|s| {
+        *s = State::from(stable_state);
+        set_certified_data(&s.root_hash());
         if let Some(set_permissions) = set_permissions {
-            s.borrow_mut().set_permissions(set_permissions);
+            s.set_permissions(set_permissions);
         }
     });
 }
 
+pub fn with_state_mut<F, R>(f: F) -> R
+where
+    F: FnOnce(&mut State) -> R,
+{
+    STATE.with(|s| f(&mut s.borrow_mut()))
+}
+
+pub fn with_state<F, R>(f: F) -> R
+where
+    F: FnOnce(&State) -> R,
+{
+    STATE.with(|s| f(&s.borrow()))
+}
+
+#[macro_export]
+macro_rules! export_canister_methods {
+    () => {
+        use $crate::asset_certification;
+        use $crate::state_machine;
+        use $crate::types;
+        use $crate::ByteBuf;
+
+        use $crate::can_commit as __ic_certified_assets_can_commit;
+        use $crate::can_prepare as __ic_certified_assets_can_prepare;
+        use $crate::is_controller as __ic_certified_assets_is_controller;
+        use $crate::is_manager_or_controller as __ic_certified_assets_is_manager_or_controller;
+
+        #[cfg(target_arch = "wasm32")]
+        #[link_section = "icp:public supported_certificate_versions"]
+        pub static CERTIFICATE_VERSIONS: [u8; 3] = $crate::SUPPORTED_CERTIFICATE_VERSIONS;
```

Comment:
We don't need the `pub` visibility here:
```suggestion
        static CERTIFICATE_VERSIONS: [u8; 3] = $crate::SUPPORTED_CERTIFICATE_VERSIONS;
```

Same for some of the methods below